src/trio/testing/_raises_group.py

```diff
@@ -0,0 +1,305 @@
+
+    from typing_extensions import TypeGuard
+
+if sys.version_info < (3, 11):
+    from exceptiongroup import BaseExceptionGroup
+
+E = TypeVar("E", bound=BaseException)
+
+
+# minimal version of pytest.ExceptionInfo in case it is not available
+@final
+class _ExceptionInfo(Generic[E]):
+    _excinfo: tuple[type[E], E, types.TracebackType] | None
+
+    def __init__(self, excinfo: tuple[type[E], E, types.TracebackType] | None):
+        self._excinfo = excinfo
+
+    def fill_unfilled(self, exc_info: tuple[type[E], E, types.TracebackType]) -> None:
+        """Fill an unfilled ExceptionInfo created with ``for_later()``."""
+        assert self._excinfo is None, "ExceptionInfo was already filled"
+        self._excinfo = exc_info
+
+    @classmethod
+    def for_later(cls) -> _ExceptionInfo[E]:
+        """Return an unfilled ExceptionInfo."""
+        return cls(None)
+
+    @property
+    def type(self) -> type[E]:
+        """The exception class."""
+        assert (
+            self._excinfo is not None
+        ), ".type can only be used after the context manager exits"
+        return self._excinfo[0]
+
+    @property
+    def value(self) -> E:
+        """The exception value."""
+        assert (
+            self._excinfo is not None
+        ), ".value can only be used after the context manager exits"
+        return self._excinfo[1]
+
+    @property
+    def tb(self) -> types.TracebackType:
+        """The exception raw traceback."""
+        assert (
+            self._excinfo is not None
+        ), ".tb can only be used after the context manager exits"
+        return self._excinfo[2]
+
+
+# this may bite users with type checkers not using pytest, but that should
+# be rare and give quite obvious errors in tests trying to do so.
+# Ignoring "incorrect import of pytest", it makes sense in this situation.
+if TYPE_CHECKING:
+    from pytest import ExceptionInfo  # noqa: PT013
+
+else:
+    try:
+        from pytest import ExceptionInfo  # noqa: PT013
+    except ImportError:  # pragma: no cover
+        ExceptionInfo = _ExceptionInfo
+
+
+# copied from pytest.ExceptionInfo
+def _stringify_exception(exc: BaseException) -> str:
+    return "\n".join(
+        [
+            str(exc),
+            *getattr(exc, "__notes__", []),
+        ]
+    )
+
+
+@final
+class Matcher(Generic[E]):
+    def __init__(
+        self,
+        exception_type: type[E] | None = None,
+        match: str | Pattern[str] | None = None,
+        check: Callable[[E], bool] | None = None,
+    ):
+        if exception_type is None and match is None and check is None:
+            raise ValueError("You must specify at least one parameter to match on.")
+        if exception_type is not None and not issubclass(exception_type, BaseException):
+            raise ValueError(
+                f"exception_type {exception_type} must be a subclass of BaseException"
+            )
+        self.exception_type = exception_type
+        self.match = match
```

Comment:
Minor and probably not necessary, but you could set `exception_type` to `BaseException` if None to avoid a condition below, and if `match` is a str do `re.compile()` so that's cached.

Reply:
the first one ended up harder to read, and might interfere with future `print-why-it-didn't-match` logic - so skipping that one. Second one sounds good though :+1: